### PR TITLE
task and project schema

### PR DESCRIPTION
**Pull request by emaness-utk**
issues #17 and #18

### CODE REVIEW

File: backend/TaskTree/Models/Project.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using Microsoft.EntityFrameworkCore;
+using System.ComponentModel.DataAnnotations;
+
+namespace TaskTree.Models
+{
+	[Index(nameof(Id), IsUnique = true)]
+	public class Project : BaseEntity
+	{
+		[Required]
+		public int Name { get; set; }
```
Comment (andlrutt) [inline backend/TaskTree/Models/Project.cs:10]: Should be a string

File: backend/TaskTree/Models/Task.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using Microsoft.EntityFrameworkCore;
+using System.ComponentModel.DataAnnotations;
+
+namespace TaskTree.Models
+{
+	[Index(nameof(Name), IsUnique = true)]
```
Comment (andlrutt) [inline backend/TaskTree/Models/Task.cs:6]: Don't think there's a need for task names to be unique, they are still distinguishable from their hierarchy and unique IDs. I could envision there being multiple tasks that have duplicate subtasks

File: backend/TaskTree/Models/Project.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿using Microsoft.EntityFrameworkCore;
+using System.ComponentModel.DataAnnotations;
+
+namespace TaskTree.Models
+{
+	[Index(nameof(Id), IsUnique = true)]
```
Comment (andlrutt) [inline backend/TaskTree/Models/Project.cs:6]: This line is not required, ID generation is handled automatically (see `BaseEntity`), so unique IDs are enforced.

**Review by andlrutt — CHANGES_REQUESTED** [review]
Good stuff! I left a few comments. And be sure to include the association between tasks and tasks, and a projects a root task.

File: backend/TaskTree/Models/Task.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+	public class Task : BaseEntity
```
Comment (andlrutt) [inline backend/TaskTree/Models/Task.cs:7]: Should have a weight field

**Review by emaness-utk — COMMENTED** [review]
these have now been completed

**Review by andlrutt — DISMISSED** [review]
Looks good as far as I can tell! Good stuff. Once we have #29 and some project API endpoints, we'll be able to test fully :)

**Comment by zaviermiller** [comment]
> `dotnet ef migrations add CreateProjectAndTask` this should usually be run as you develop the entities that way we can easily roll back different parts as need be but this isn't a large changing project so having them both created in one migration should be fine.

tried doing this, the sql schema it builds isnt quite right. im investigating now to figure out why. here are the issues im seeing:
- Project and Task table should be called Projects and Tasks, small but makes it consistent
- no UserId in Project
- no ProjectId in Task

**Comment by emaness-utk** [comment]
Are these issues with the models or the SQL? I was under the impression that visibility would only go downward (i.e. User sees project, but not the other way, and same with Project to Task) to keep things compartmentalized.

**Comment by zaviermiller** [comment]
Ok, so I've done a bit more research and have figured out the EF relation stuff. Check out the entities in the `Models` folder to make sure you guys think that will be good for use.

Here are the tables in MySQL now:

```
mysql> DESCRIBE Users;
+-----------+--------------+------+-----+---------+----------------+
| Field     | Type         | Null | Key | Default | Extra          |
+-----------+--------------+------+-----+---------+----------------+
| Id        | bigint       | NO   | PRI | NULL    | auto_increment |
| Username  | varchar(255) | NO   | UNI | NULL    |                |
| Password  | longtext     | NO   |     | NULL    |                |
| FirstName | longtext     | YES  |     | NULL    |                |
| LastName  | longtext     | YES  |     | NULL    |                |
| CreatedAt | datetime(6)  | NO   |     | NULL    |                |
| UpdatedAt | datetime(6)  | NO   |     | NULL    |                |
+-----------+--------------+------+-----+---------+----------------+
7 rows in set (0.01 sec)

mysql> DESCRIBE Projects;
+-------------+-------------+------+-----+---------+----------------+
| Field       | Type        | Null | Key | Default | Extra          |
+-------------+-------------+------+-----+---------+----------------+
| Id          | bigint      | NO   | PRI | NULL    | auto_increment |
| Name        | longtext    | NO   |     | NULL    |                |
| Description | longtext    | YES  |     | NULL    |                |
| Progress    | double      | YES  |     | NULL    |                |
| CreatedAt   | datetime(6) | NO   |     | NULL    |                |
| UpdatedAt   | datetime(6) | NO   |     | NULL    |                |
| UserId      | bigint      | NO   | MUL | 0       |                |
+-------------+-------------+------+-----+---------+----------------+
7 rows in set (0.00 sec)

mysql> DESCRIBE Tasks;
+-------------+-------------+------+-----+---------+----------------+
| Field       | Type        | Null | Key | Default | Extra          |
+-------------+-------------+------+-----+---------+----------------+
| Id          | bigint      | NO   | PRI | NULL    | auto_increment |
| Name        | longtext    | NO   |     | NULL    |                |
| Description | longtext    | YES  |     | NULL    |                |
| Progress    | double      | YES  |     | NULL    |                |
| Weight      | double      | YES  |     | NULL    |                |
| CompletedAt | datetime(6) | YES  |     | NULL    |                |
| CreatedAt   | datetime(6) | NO   |     | NULL    |                |
| UpdatedAt   | datetime(6) | NO   |     | NULL    |                |
| ProjectId   | bigint      | YES  | UNI | NULL    |                |
| ParentId    | bigint      | YES  | MUL | NULL    |                |
+-------------+-------------+------+-----+---------+----------------+
10 rows in set (0.01 sec)
```

The tasks have a project id and not the other way around to signify that the task is *dependent* on the project, meaning that if a project is deleted the root task will be deleted instead of the other way around.

another thing of note is that tasks do not require a project nor a parent task, so we need to be careful that we dont have lone tasks. this shouldnt be a big deal but if you guys think one of those should be required, we should discuss.

**Comment by zaviermiller** [comment]
> Are these issues with the models or the SQL? I was under the impression that visibility would only go downward (i.e. User sees project, but not the other way, and same with Project to Task) to keep things compartmentalized.

also just to clear up, i was just referring to how it is actually stored in the database. since SQL doesnt store arrays, you need to go bottom up as opposed to the inuitive top down

**Review by andlrutt — APPROVED** [review]
Looks good as far as I can tell

**Review by zaviermiller — APPROVED** [review]

**Pull request merged**